File: lib/http_helpers.rb
require 'json'

module HttpHelpers
  module_function

  def json(response, status, payload)
    response.status = status
    response['Content-Type'] = 'application/json'
    response.body = JSON.generate(payload)
  end

  def parse_json(request)
    body = request.body.to_s
    body.empty? ? {} : JSON.parse(body)
  rescue JSON::ParserError
    raise DomainError.new('invalid JSON body', status: 400)
  end
end

File: src/http_helpers.py
import json
from urllib.parse import parse_qs, urlparse


def read_json(handler):
    length = int(handler.headers.get("Content-Length", "0"))
    raw = handler.rfile.read(length) if length else b"{}"
    return json.loads(raw.decode("utf-8"))


def query_params(path):
    parsed = urlparse(path)
    return parsed.path, {key: values[-1] for key, values in parse_qs(parsed.query).items()}


def send_json(handler, status, payload):
    body = json.dumps(payload).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")
    handler.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
    handler.send_header("Content-Length", str(len(body)))
    handler.end_headers()
    handler.wfile.write(body)
